### PR TITLE
Report unreachable on enums

### LINKED ISSUE #58379: Missing reachability error on non-const enum
### 🔎 Search Terms

non-const enum unreachable error missing

### 🕗 Version & Regression Information

Nightly

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAIwDYEMAWAKAlIgbwFgAoASHQDFwI8BuU0siBAZzlQFMA6VOAc2wAhdACduAQVwNyLMOy68B2MJwDuiSnDh5cTOQp59BIgF7cAbulQhO0xiSajOUEKLAymoSLASIqNHiETMxsHEbKIuJSMmSh8uFKgqoaWjq4euTxhknC6OZWNnaxAL4OZJxgIAC2iFGEiBKIZY6yGKysmtoNAB6IALyIAIwATADMACzN5WDo1ZysAA7oEJx1+Q2cPYtwolCIBvuFtgPD41MtJUA

### 💻 Code

```ts
function blah() {
	aFunc();

	console.log(Bar.A);
	console.log(new Foo())
	console.log(Baz.value);


	return;

	function aFunc() {
		console.log(Bar.A);
		console.log(new Foo())
		console.log(Baz.value);
	}

	enum Bar { A }

	class Foo { x = 1234 }

	namespace Baz { export const value = 1234 }
}
```

### 🙁 Actual behavior

No error on `Bar`.

### 🙂 Expected behavior

Error on `Bar` - its' definition will never run.

### Additional information about the issue

This is easy to accidentally cause in our own codebase, and `esbuild` actually covers up the problem by hoisting/inlining the enum - the problem only shows up in an unbundled build.

`checkUnreachable` in `binder.ts` is missing a check for `EnumDeclaration`s (and a bail from that check for `const enum`s).

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #58379

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot test it
@typescript-bot pack this

**Comment by jakebailey** [comment]
Even this is not enough: https://www.typescriptlang.org/play/?ts=5.5.0-pr-58380-2#code/GYVwdgxgLglg9mABAIwDYEMAWAKAlIgbwFgAoASHQDFwI8BuU0siBAZzlQFMA6VOAc2wAhdACduAQVwNyLMOy68B2NFm4AHUXHXSmchTz6CwnAO6JKcOHlx62HQ8pEAvbgDd0qEJ10lG5UU4oEFEwGSZQSFgERCoaPEImZntFI2ExSV8yZPkHJUFVTA0tHRls-Ty0k3NLa1xbchyDfPTXDy8fMoBffyaoFAxMRABeQkRNbQAuRABGACYAZgAWRC7w8k4wEABbRBFRMYlV3ogMVlYLKzGADxHZxZWev3IwdG3OVnV0CE499Gcxpxruo4KJ+vp+u1vHd5stjiQukA

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -3785,6 +3785,7 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
                 (isStatementButNotDeclaration(node) && node.kind !== SyntaxKind.EmptyStatement) ||
                 // report error on class declarations
                 node.kind === SyntaxKind.ClassDeclaration ||
+                (node.kind === SyntaxKind.EnumDeclaration && (!isEnumConst(node as EnumDeclaration) || shouldPreserveConstEnums(options))) ||
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts:3788]:
```suggestion
                // report errors on enums with preserved emit
                (node.kind === SyntaxKind.EnumDeclaration && (!isEnumConst(node as EnumDeclaration) || shouldPreserveConstEnums(options))) ||

```

**Review by DanielRosenwasser — APPROVED** [review]
Not a great error span but it's fine.

**Comment by jakebailey** [comment]
Yeah, this isn't quite done; classes get a better message here.

**Comment by jakebailey** [comment]
I'm actually confused what I meant by "classes have a better message"; I updated the test and it seems like they're identical.

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -3789,6 +3789,9 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
                 (isStatementButNotDeclaration(node) && node.kind !== SyntaxKind.EmptyStatement) ||
                 // report error on class declarations
                 node.kind === SyntaxKind.ClassDeclaration ||
+                // report errors on enums with preserved emit
+                (node.kind === SyntaxKind.EnumDeclaration && (!isEnumConst(node as EnumDeclaration) || shouldPreserveConstEnums(options))) ||
+
```
Comment (jakebailey) [inline src/compiler/binder.ts:3794]:
```suggestion
```

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -3789,6 +3789,8 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
                 (isStatementButNotDeclaration(node) && node.kind !== SyntaxKind.EmptyStatement) ||
                 // report error on class declarations
                 node.kind === SyntaxKind.ClassDeclaration ||
+                // report errors on enums with preserved emit
+                (node.kind === SyntaxKind.EnumDeclaration && (!isEnumConst(node as EnumDeclaration) || shouldPreserveConstEnums(options))) ||
                 // report error on instantiated modules or const-enums only modules if preserveConstEnums is set
```
Comment (jakebailey) [inline src/compiler/binder.ts:3794]:
```suggestion
                // report error on instantiated modules
```

**Comment by jakebailey** [comment]
Oh, https://github.com/microsoft/TypeScript/pull/58380#issuecomment-2087750290 wasn't fixed by this.

File: tests/baselines/reference/unreachableDeclarations(preserveconstenums=false).errors.txt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    		console.log(blah.prop);
+    		console.log(new Foo())
+    		console.log(Baz.value);
+    	}
+    
+    	const blah = { prop: 1234 };
+    	~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS7027: Unreachable code detected.
+    
+    	enum Bar { A }
```
Comment (jakebailey) [inline tests/baselines/reference/unreachableDeclarations(preserveconstenums=false).errors.txt:124]: Problem here is that there is no error on this yet.

**Review by weswigham — APPROVED** [review]

**Comment by jakebailey** [comment]
@typescript-bot test it

**Pull request merged**